.travis.yml

```diff
@@ -4,37 +4,46 @@ language: generic
 sudo: required
 
 env:
-  DOCKER_COMPOSE_VERSION: 1.16.1
+  global:
+    - DOCKER_COMPOSE_VERSION=1.26.2
+    - ANSIBLE_VERSION=2.9.9
+  jobs:
+    - PLATFORM=centos PLAYBOOK=user.yml DOCKER_CACHE_DIR=$HOME/docker/$PLATFORM
+    - PLATFORM=debian PLAYBOOK=user.yml DOCKER_CACHE_DIR=$HOME/docker/$PLATFORM
+    - PLATFORM=ubuntu PLAYBOOK=user.yml DOCKER_CACHE_DIR=$HOME/docker/$PLATFORM
+    - PLATFORM=centos PLAYBOOK=root.yml DOCKER_CACHE_DIR=$HOME/docker/$PLATFORM
+    - PLATFORM=debian PLAYBOOK=root.yml DOCKER_CACHE_DIR=$HOME/docker/$PLATFORM
+    - PLATFORM=ubuntu PLAYBOOK=root.yml DOCKER_CACHE_DIR=$HOME/docker/$PLATFORM
 
 services:
   - docker
 
 cache:
   pip: true
   directories:
-    - $HOME/docker
+    - $DOCKER_CACHE_DIR
 
 before_install:
-  - if [[ -d $HOME/docker ]]; then ls $HOME/docker/*.tar.gz | xargs -I {file} sh -c "zcat {file} | docker load"; fi
+  - if [[ -d $DOCKER_CACHE_DIR ]]; then ls $DOCKER_CACHE_DIR/*.tar.gz | xargs -I {file} sh -c "zcat {file} | docker load"; fi
   - sudo rm /usr/local/bin/docker-compose
   - curl -L https://github.com/docker/compose/releases/download/$DOCKER_COMPOSE_VERSION/docker-compose-`uname -s`-`uname -m` > docker-compose
   - chmod +x docker-compose
   - sudo mv docker-compose /usr/local/bin
 
 install:
-  - sudo pip install ansible==2.4.6.0 > /dev/null 2>&1
+  - sudo pip install ansible==$ANSIBLE_VERSION > /dev/null 2>&1
 
 script:
   - cd tests
-  - docker-compose up -d
-  - ansible-playbook user.yml -l centos
-  - ansible-playbook user.yml -l debian
-  - ansible-playbook user.yml -l ubuntu
+  - docker-compose config --services | grep $PLATFORM | xargs docker-compose up -d
+  - ansible-playbook $PLAYBOOK -l $PLATFORM
+  - docker-compose down
```

Comment:
I wonder what will happen if the previous command fails. Will the VM still shut-down? I presume yes, it is more curiosity as I didn't test that type of scenario on Travis.

Reply:
AFAIK, Travis CI doesn't behave in a "fail-fast" fashion by default, i.e. every statement will be executed (which might not be the desired behaviour, and I'll propose a change on that regard).

See https://github.com/travis-ci/travis-ci/issues/1066

PS: In this very specific example, the `docker-compose down` could (and maybe should) simply be removed as it adds nothing, and slows the build for nothing (it was initially there, when the build "matrix" was still tested within a single sequential job).

Reply:
@thbar I propose to try "fail fast" (see 943ff233833892bf97be6b149fce09cdd9007f67).

Note: not 100% sure, this will be enough to be applied on `script:` stages as well... (to be verified "soon" ;-)